README.md
# Grafana Sequential Data Store Sample

| :loudspeaker: **Notice**: This sample has been Archived. Dependencies will not be updated and pipelines will not be run. Please contact samples@osisoft.com or visit The OSIsoft Feedback Site for comments |
| -----------------------------------------------------------------------------------------------|

**Version:** ARCHIVED

[![Build Status](https://dev.azure.com/osieng/engineering/_apis/build/status/product-readiness/ADH/aveva.sample-adh-grafana-nodejs?branchName=main)](https://dev.azure.com/osieng/engineering/_build/latest?definitionId=2619&branchName=main)

This sample demonstrates how to build a [Grafana](https://grafana.com/) plugin that runs queries against the Sequential Data Store of AVEVA Data Hub or Edge Data Store. The sample performs normal "Get Values" calls against a specified stream in SDS, using the time range of the Grafana dashboard. See the [Grafana Documentation](https://grafana.com/docs/grafana/latest/developers/plugins/) for more information on developing Grafana plugins.

## Requirements

- [Grafana 7.0+](https://grafana.com/grafana/download)
- Web Browser with JavaScript enabled
- [NodeJS](https://nodejs.org/en/)
- [Git](https://git-scm.com/download/win)
- If using AVEVA Data Hub, register a Client Credentials Client in AVEVA Data Hub; a client secret will need to be provided to the sample plugin configuration
- If using Edge Data Store, the browser must be running local to a running copy of Edge Data Store

## Running the Sample with installed Grafana

1. Copy this folder to your Grafana server's plugins directory, like `.../grafana/data/plugins`
1. (Optional) If using other plugins, rename the folder to `aveva-data-hub-sample`
1. Open a command prompt inside that folder
1. Install dependencies, using `npm ci`
1. Build the plugin, using `npm run build` (or `npm run dev` for browser debugging)
1. Restart the Grafana server to load the new plugin
1. Open the Grafana configuration and set the parameter `allow_loading_unsigned_plugins` equal to `aveva-sds-sample` or to the name of the folder set in step 2 (see [Grafana docs](https://grafana.com/docs/grafana/latest/administration/configuration/#allow_loading_unsigned_plugins))
1. Add a new Grafana datasource using the sample (see [Grafana docs](https://grafana.com/docs/grafana/latest/features/datasources/add-a-data-source/))
1. Choose whether to query against AVEVA Data Hub or Edge Data Store
1. Enter the relevant required information; if using ADH, the client secret will be encrypted in the Grafana server and HTTP requests to ADH will be made by a server-side proxy, as described in the [Grafana docs](https://grafana.com/docs/grafana/latest/developers/plugins/authentication/)
1. Open a new or existing Grafana dashboard, and choose the Sequential Data Store Sample as the data source
1. Enter your Namespace (if querying ADH) and Stream, and data will populate into the dashboard from the stream for the dashboard's time range

## Running the Sample with Docker

1. Open a command prompt inside this folder
1. Build the container using `docker build -t grafana-adh .`  
   _Note: The dockerfile being built contains an ENV statement that creates an [environment variable](https://grafana.com/docs/grafana/latest/administration/configuration/#configure-with-environment-variables) that overrides an option in the grafana config. In this case, the `allow_loading_unsigned_plugins` option is being overridden to allow the [unsigned plugin](https://grafana.com/docs/grafana/latest/administration/configuration/#allow_loading_unsigned_plugins) in this sample to be used._
1. Run the container using `docker run -d --name=grafana -p 3000:3000 grafana-adh`
1. Navigate to localhost:3000 to configure data sources and view data

## Using ADH OAuth login to Grafana

A limitation of this sample is that the use of a Client Credentials Client means that anyone using the Data Source in Grafana receives the same access inside AVEVA Data Hub. Grafana supports generic OAuth login to the Grafana server, and can forward its token to the destination data source, including ADH. See [here](https://grafana.com/docs/grafana/latest/auth/generic-oauth/) for more information on this feature of Grafana. This feature could be used to ensure individual users must log in to receive the appropriate permissions in ADH.

However, Grafana's implementation of the Authorization Code Flow does not use [Proof of Key Code Exchange](https://oauth.net/2/pkce/), or PKCE, which is an additional security layer required by ADH Authorization Code Clients. Grafana also does not support ADH Hybrid Clients as it does not support a `POST` back from the authentication server, nor does it support the `response_type` and `response_mode` headers required by that flow.

If you are interested in using using ADH OAuth in your Grafana server, please check and comment on [this issue](https://github.com/grafana/grafana/issues/26350) in the Grafana GitHub repository.

## Using Community Data

1. Add a new Grafana datasource using the sample (see [Grafana docs](https://grafana.com/docs/grafana/latest/features/datasources/add-a-data-source/))
1. Choose AVEVA Data Hub
1. Toggle the "Community Data" switch to 'true'
1. Enter the relevant required information. You can find the Community ID in the URL of the Community Details page.

## Running the Automated Tests

1. Open a command prompt inside this folder
1. Install dependencies, using `npm ci`
1. Run the tests, using `npm test`

---

For the main ADH page [ReadMe](https://github.com/osisoft/OSI-Samples-OCS)  
For the main samples page [ReadMe](https://github.com/osisoft/OSI-Samples)


HISTORY.md
# Version History

## ARCHIVED / 2022-09-02

- A new version of the grafana sample that introduces new features and uses a new framework has been released
- The new sample can be found in [sample-adh-grafana_backend_plugin-datasource](https://github.com/osisoft/sample-adh-grafana_backend_plugin-datasource)

## 1.4.9 / 2022-08-25

- Resolve possible vulnerability by not running docker container as root

## 1.4.8 / 2022-08-18

- Add missing dependency for docker build

## 1.4.7 / 2022-08-17

- Automate dependabot approval and automerge

## 1.4.6 / 2022-08-09

- Add dependabot.yml to automate dependency checks
- Remove weekly tests
- Parameterize pool and demands

## 1.4.5 / 2022-07-13

- Update spyOn usage to be compatible with @grafana/toolkit versions >= 9.0.0

## 1.4.4 / 2022-06-20

- Updated dependencies

## 1.4.3 / 2022-05-23

- Update communities queries to be compliant with changes made to the API

## 1.4.2 / 2022-05-13

- Resolve linting issues
- Resolve React component issues

## 1.4.1 / 2022-05-10

- Updated dependencies

## 1.4.0 / 2022-04-20

- Add debouncing in query editor to reduce number of queries to ADH

## 1.3.1 / 2021-03-15

- Updated dependencies

## 1.3.0 / 2022-01-27

- Updated for AVEVA Data Hub

## 1.2.9 / 2022-01-06

- Update pipelines to reference internal analysis templates

## 1.2.8 / 2021-12-10

- Translate boolean stream vales to numerical

## 1.2.7 / 2021-12-09

- Change stream retrieval to show stream names instead of Ids
- Have stream names show in trend legends as opposed to Ids
- Add access parameter

## 1.2.6 / 2021-10-21

- Updated dependencies

## 1.2.5 / 2021-09-07

- Updated dependencies

## 1.2.4 / 2021-08-11

- Updated dependencies
- Cleaned up gitignore

## 1.2.3 / 2021-07-20

- Updated dependencies

## 1.2.2 / 2021-07-09

- Allow the ocs-sds-sample plugin as an unsigned plugin

## 1.2.1 / 2021-07-06

- Updated dependencies

## 1.2.0 / 2021-06-01

- Added the ability to view data from OCS communities
- Added a dockerfile for containerized sample with Grafana server

## 1.1.7 / 2021-05-24

- Updated dependencies

## 1.1.6 / 2021-04-13

- Updated dependencies

## 1.1.5 / 2021-03-02

- Updated pipeline to use internal agent pool
- Updated pipeline to reference main branch

## 1.1.4 / 2021-01-28

- Fix linting issues

## 1.1.3 / 2021-01-26

- Updated dependencies

## 1.1.2 / 2020-11-02

- Updated dependencies

## 1.1.1 / 2020-09-14

- Updated dependencies

## 1.1.0 / 2020-08-11

- Converted HTML partials to React components
- Added namespace to Data Source configuration
- Added selection field for stream ID to query editor

## 1.0.2 / 2020-08-06

- Updated dependencies

## 1.0.1 / 2020-06-11

- Updated dependencies

## 1.0.0 / 2020-06-02

- Initial public release


src/DataSource.spec.ts
import { DataSourceInstanceSettings, MutableDataFrame, FieldType } from '@grafana/data';

import { SdsDataSource } from 'DataSource';
import { SdsDataSourceOptions, SdsDataSourceType } from 'types';

describe('SdsDataSource', () => {
  const url = 'URL';
  const edsPort = 'PORT';
  const adhUrl = 'URL';
  const adhVersion = 'VERSION';
  const adhTenant = 'TENANT';
  const adhClient = 'CLIENT';
  const oauthPassThru = false;
  const namespace = 'NAMESPACE';
  const adhUseCommunity = false;
  const adhCommunity = 'COMMUNITY';
  const adhSettings: DataSourceInstanceSettings<SdsDataSourceOptions> = {
    id: 0,
    uid: '',
    name: '',
    access: 'direct',
    type: '',
    url,
    meta: null as any,
    jsonData: {
      type: SdsDataSourceType.ADH,
      edsPort: edsPort,
      adhUrl: adhUrl,
      adhVersion: adhVersion,
      adhTenant: adhTenant,
      adhClient: adhClient,
      oauthPassThru,
      namespace,
      adhUseCommunity: adhUseCommunity,
      adhCommunity: adhCommunity,
    },
  };
  const adhCommSettings = { ...adhSettings, ...{ jsonData: { ...adhSettings.jsonData, adhUseCommunity: true } } };
  const edsSettings = { ...adhSettings, ...{ jsonData: { ...adhSettings.jsonData, type: SdsDataSourceType.EDS } } };
  const backendSrv = {
    datasourceRequest: () => new Promise((r) => r),
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('constructor', () => {
    it('should use passed in data source information', () => {
      const datasource = new SdsDataSource(adhSettings, backendSrv as any);
      expect(datasource.proxyUrl).toEqual(url);
      expect(datasource.type).toEqual(SdsDataSourceType.ADH);
      expect(datasource.edsPort).toEqual(edsPort);
      expect(datasource.adhUrl).toEqual(adhUrl);
      expect(datasource.adhVersion).toEqual(adhVersion);
      expect(datasource.adhTenant).toEqual(adhTenant);
      expect(datasource.adhUseCommunity).toEqual(adhUseCommunity);
      expect(datasource.adhCommunity).toEqual(adhCommunity);
      expect(datasource.oauthPassThru).toEqual(oauthPassThru);
      expect(datasource.namespace).toEqual(namespace);
    });

    it('should handle empty jsonData', () => {
      const nullSettings: any = { ...adhSettings, ...{ url: null, jsonData: null } };
      const datasource = new SdsDataSource(nullSettings, backendSrv as any);
      expect(datasource.proxyUrl).toEqual('');
      expect(datasource.type).toEqual(SdsDataSourceType.ADH);
      expect(datasource.edsPort).toEqual('5590');
      expect(datasource.adhUrl).toEqual('');
      expect(datasource.adhVersion).toEqual('v1');
      expect(datasource.adhTenant).toEqual('');
      expect(datasource.adhUseCommunity).toEqual(false);
      expect(datasource.adhCommunity).toEqual('');
      expect(datasource.oauthPassThru).toEqual(false);
      expect(datasource.namespace).toEqual('');
    });
  });

  describe('getStreamsUrl', () => {
    it('should return the correct URL for ADH', () => {
      const datasource = new SdsDataSource(adhSettings, backendSrv as any);
      expect(datasource.streamsUrl).toEqual('URL/adh/api/VERSION/tenants/TENANT/namespaces/NAMESPACE/streams');
    });

    it('should return the correct URL for ADH communities', () => {
      const datasource = new SdsDataSource(adhCommSettings, backendSrv as any);
      expect(datasource.streamsUrl).toEqual('URL/community/api/VERSION/search/communities/COMMUNITY/streams');
    });

    it('should return the correct URL for EDS', () => {
      const datasource = new SdsDataSource(edsSettings, backendSrv as any);
      expect(datasource.streamsUrl).toEqual(
        'http://localhost:PORT/api/v1/tenants/default/namespaces/NAMESPACE/streams'
      );
    });
  });

  describe('query', () => {
    it('should query with the expected parameters', (done) => {
      jest.spyOn(backendSrv, 'datasourceRequest').mockReturnValue(
        Promise.resolve({
          data: [
            {
              TimeStamp: '2020-01-01',
              Boolean: true,
              Number: 1,
              String: 'A',
            },
          ],
        })
      );
      const datasource = new SdsDataSource(adhSettings, backendSrv as any);
      const options = {
        range: {
          from: {
            utc: () => ({
              format: () => 'FROM',
            }),
          },
          to: {
            utc: () => ({
              format: () => 'TO',
            }),
          },
        },
        targets: [
          {
            refId: 'REFID',
            namespace: 'NAMESPACE',
            streamId: 'STREAM',
            streamName: 'STREAM',
          },
        ],
      };
      const response = datasource.query(options as any);
      expect(backendSrv.datasourceRequest).toHaveBeenCalledWith({
        url: 'URL/adh/api/VERSION/tenants/TENANT/namespaces/NAMESPACE/streams/STREAM/data?startIndex=FROM&endIndex=TO',
        method: 'GET',
      });
      response.then((r) => {
        expect(JSON.stringify(r)).toEqual(
          JSON.stringify({
            data: [
              new MutableDataFrame({
                refId: 'REFID',
                name: 'STREAM',
                fields: [
                  {
                    name: 'TimeStamp',
                    type: FieldType.time,
                    values: [Date.parse('2020-01-01')],
                  },
                  {
                    name: 'Boolean',
                    type: FieldType.number,
                    values: [1],
                  },
                  {
                    name: 'Number',
                    type: FieldType.number,
                    values: [1],
                  },
                  {
                    name: 'String',
                    type: FieldType.string,
                    values: ['A'],
                  },
                ],
              }),
            ],
          })
        );
        done();
      });
    });

    it('should query a community with the expected parameters', (done) => {
      jest.spyOn(backendSrv, 'datasourceRequest').mockReturnValue(
        Promise.resolve({
          data: [
            {
              TimeStamp: '2020-01-01',
              Boolean: true,
              Number: 1,
              String: 'A',
            },
          ],
        })
      );
      const datasource = new SdsDataSource(adhCommSettings, backendSrv as any);
      const options = {
        range: {
          from: {
            utc: () => ({
              format: () => 'FROM',
            }),
          },
          to: {
            utc: () => ({
              format: () => 'TO',
            }),
          },
        },
        targets: [
          {
            refId: 'REFID',
            namespace: 'NAMESPACE',
            streamId: 'https://uswe.datahub.connect.aveva.com/streampath',
            streamName: 'STREAM',
          },
        ],
      };
      const response = datasource.query(options as any);
      expect(backendSrv.datasourceRequest).toHaveBeenCalledWith({
        url: 'URL/community/streampath/data?startIndex=FROM&endIndex=TO',
        method: 'GET',
      });
      response.then((r) => {
        expect(JSON.stringify(r)).toEqual(
          JSON.stringify({
            data: [
              new MutableDataFrame({
                refId: 'REFID',
                name: 'STREAM',
                fields: [
                  {
                    name: 'TimeStamp',
                    type: FieldType.time,
                    values: [Date.parse('2020-01-01')],
                  },
                  {
                    name: 'Boolean',
                    type: FieldType.number,
                    values: [1],
                  },
                  {
                    name: 'Number',
                    type: FieldType.number,
                    values: [1],
                  },
                  {
                    name: 'String',
                    type: FieldType.string,
                    values: ['A'],
                  },
                ],
              }),
            ],
          })
        );
        done();
      });
    });

    it('should handle invalid query', (done) => {
      jest.spyOn(backendSrv, 'datasourceRequest');
      const datasource = new SdsDataSource(adhSettings, backendSrv as any);
      const options = {
        range: null,
        targets: [
          {
            refId: 'REFID',
            namespace: 'NAMESPACE',
            streamId: null,
            streamName: 'STREAM',
          },
        ],
      };
      const response = datasource.query(options as any);
      expect(backendSrv.datasourceRequest).not.toHaveBeenCalled();
      response.then((r) => {
        expect(JSON.stringify(r)).toEqual(JSON.stringify({ data: [new MutableDataFrame()] }));
        done();
      });
    });
  });

  describe('getStreams', () => {
    it('should return empty if namespace is not defined', (done) => {
      const datasource = new SdsDataSource(adhSettings, backendSrv as any);
      datasource.namespace = '';
      const result = datasource.getStreams('');
      result.then((r) => {
        expect(r).toEqual([]);
        done();
      });
    });

    it('should query for streams', (done) => {
      const Id = 'Stream';
      const Name = 'Stream';
      jest.spyOn(backendSrv, 'datasourceRequest').mockReturnValue(Promise.resolve({ data: [{ Id, Name }] }));
      const datasource = new SdsDataSource(adhSettings, backendSrv as any);
      const result = datasource.getStreams('test');
      result.then((r) => {
        expect(r).toEqual([{ value: Id, label: Name }]);
        done();
      });
    });

    it('should query a community for streams', (done) => {
      const Id = 'Stream';
      const Self = 'Self';
      const Name = 'Stream';
      jest.spyOn(backendSrv, 'datasourceRequest').mockReturnValue(Promise.resolve({ data: [{ Self, Id, Name }] }));
      const datasource = new SdsDataSource(adhCommSettings, backendSrv as any);
      const result = datasource.getStreams('test');
      result.then((r) => {
        expect(r).toEqual([{ value: Self, label: Name }]);
        done();
      });
    });
  });

  describe('testDatasource', () => {
    it('should run a test query', (done) => {
      jest.spyOn(backendSrv, 'datasourceRequest').mockReturnValue(
        Promise.resolve({
          status: 200,
        })
      );
      const datasource = new SdsDataSource(adhSettings, backendSrv as any);
      const response = datasource.testDatasource();
      expect(backendSrv.datasourceRequest).toHaveBeenCalledWith({
        url: 'URL/adh/api/VERSION/tenants/TENANT/namespaces/NAMESPACE/streams',
        method: 'GET',
      });
      response.then((r) => {
        expect(r).toEqual({
          status: 'success',
          message: 'Data source is working',
        });
        done();
      });
    });

    it('should handle test failure', (done) => {
      jest.spyOn(backendSrv, 'datasourceRequest').mockReturnValue(
        Promise.resolve({
          status: 400,
          statusText: 'Error',
        })
      );
      const datasource = new SdsDataSource(adhSettings, backendSrv as any);
      const response = datasource.testDatasource();
      expect(backendSrv.datasourceRequest).toHaveBeenCalledWith({
        url: 'URL/adh/api/VERSION/tenants/TENANT/namespaces/NAMESPACE/streams',
        method: 'GET',
      });
      response.then((r) => {
        expect(r).toEqual({
          status: 'error',
          message: '400: Error',
        });
        done();
      });
    });
  });
});


src/DataSource.ts
import {
  DataQueryRequest,
  DataQueryResponse,
  DataSourceApi,
  DataSourceInstanceSettings,
  MutableDataFrame,
  FieldType,
  SelectableValue,
} from '@grafana/data';

import { SdsQuery, SdsDataSourceOptions, SdsDataSourceType } from 'types';

export declare type BackendSrvRequest = {
  url: string;
  method?: string;
};

export interface BackendSrv {
  datasourceRequest(options: BackendSrvRequest): Promise<any>;
}

export class SdsDataSource extends DataSourceApi<SdsQuery, SdsDataSourceOptions> {
  name: string;
  proxyUrl: string;

  type: SdsDataSourceType;
  edsPort: string;
  adhUrl: string;
  adhVersion: string;
  adhTenant: string;
  adhUseCommunity: boolean;
  adhCommunity: string;
  oauthPassThru: boolean;
  namespace: string;

  get streamsUrl() {
    return this.type === SdsDataSourceType.ADH
      ? this.adhUseCommunity === true
        ? `${this.proxyUrl}/community/api/${this.adhVersion}/search/communities/${this.adhCommunity}/streams`
        : `${this.proxyUrl}/adh/api/${this.adhVersion}/tenants/${this.adhTenant}/namespaces/${this.namespace}/streams`
      : `http://localhost:${this.edsPort}/api/v1/tenants/default/namespaces/${this.namespace}/streams`;
  }

  /** @ngInject */
  constructor(instanceSettings: DataSourceInstanceSettings<SdsDataSourceOptions>, private backendSrv: BackendSrv) {
    super(instanceSettings);
    this.name = instanceSettings.name;
    this.proxyUrl = instanceSettings.url ? instanceSettings.url.trim() : '';
    this.backendSrv = backendSrv;

    this.type = instanceSettings.jsonData?.type || SdsDataSourceType.ADH;
    this.edsPort = instanceSettings.jsonData?.edsPort || '5590';
    this.adhUrl = instanceSettings.jsonData?.adhUrl || '';
    this.adhVersion = instanceSettings.jsonData?.adhVersion || 'v1';
    this.adhTenant = instanceSettings.jsonData?.adhTenant || '';
    this.adhUseCommunity = instanceSettings.jsonData?.adhUseCommunity || false;
    this.adhCommunity = instanceSettings.jsonData?.adhCommunity || '';
    this.oauthPassThru = instanceSettings.jsonData?.oauthPassThru || false;
    this.namespace = instanceSettings.jsonData?.namespace || '';
  }

  async query(options: DataQueryRequest<SdsQuery>): Promise<DataQueryResponse> {
    const from = options.range?.from.utc().format();
    const to = options.range?.to.utc().format();
    const requests = options.targets.map((target) => {
      if (!target.streamId) {
        return new Promise((resolve) => resolve(null));
      }
      if (this.adhUseCommunity) {
        const url = new URL(target.streamId);

        return this.backendSrv.datasourceRequest({
          url: `${this.proxyUrl}/community${url.pathname}/data?startIndex=${from}&endIndex=${to}`,
          method: 'GET',
        });
      } else {
        return this.backendSrv.datasourceRequest({
          url: `${this.streamsUrl}/${target.streamId}/data?startIndex=${from}&endIndex=${to}`,
          method: 'GET',
        });
      }
    });

    const data = await Promise.all(requests).then((responses) => {
      let i = 0;
      return responses.map((r: any) => {
        if (!r || !r.data.length) {
          return new MutableDataFrame();
        }

        const target = options.targets[i];
        i++;
        return new MutableDataFrame({
          refId: target.refId,
          name: target.streamName,
          fields: Object.keys(r.data[0]).map((name) => {
            const val0 = r.data[0][name];
            const date = Date.parse(val0);
            const num = Number(val0);
            const type =
              typeof val0 === 'string' && !isNaN(date)
                ? FieldType.time
                : val0 === true || val0 === false
                ? FieldType.boolean
                : !isNaN(num)
                ? FieldType.number
                : FieldType.string;
            let values = [];
            if (type === FieldType.boolean) {
              values = r.data.map((d) => {
                return d[name]?.toString().toLowerCase() === 'true' ? 1 : 0;
              });
            } else {
              values = r.data.map((d) => (type === FieldType.time ? Date.parse(d[name]) : d[name]));
            }

            return {
              name,
              values: values,
              type: type === FieldType.boolean ? FieldType.number : type,
            };
          }),
        });
      });
    });

    return { data };
  }

  async getStreams(query: string): Promise<Array<SelectableValue<string>>> {
    const url = query ? `${this.streamsUrl}?query=*${query}*` : this.streamsUrl;
    const requests = this.backendSrv.datasourceRequest({ url, method: 'GET' });
    if (this.adhUseCommunity === true) {
      return await Promise.resolve(requests).then((responses) =>
        Object.keys(responses.data).map((r) => ({ value: responses.data[r].Self, label: responses.data[r].Name }))
      );
    } else if (this.namespace) {
      return await Promise.resolve(requests).then((responses) =>
        Object.keys(responses.data).map((r) => ({ value: responses.data[r].Id, label: responses.data[r].Name }))
      );
    } else {
      return await new Promise((resolve) => resolve([]));
    }
  }

  async testDatasource() {
    return this.backendSrv
      .datasourceRequest({
        url: this.streamsUrl,
        method: 'GET',
      })
      .then((r) => {
        if (r.status === 200) {
          return {
            status: 'success',
            message: 'Data source is working',
          };
        } else {
          return {
            status: 'error',
            message: `${r.status}: ${r.statusText}`,
          };
        }
      });
  }
}
